File: 2290-minimum-obstacle-removal-to-reach-corner/2290-minimum-obstacle-removal-to-reach-corner.py
class Solution:
    def minimumObstacles(self, grid: List[List[int]]) -> int:
        def is_valid(x: int, y: int) -> bool:
            return 0 <= x < m and 0 <= y < n
        
        m, n = len(grid), len(grid[0])
        visited = [[float('inf')] * n for _ in range(m)]
        pq = [(0, 0, 0)]
        
        while pq:
            min_remove, i, j = heappop(pq)
            
            if min_remove >= visited[i][j]:
                continue
            
            if i == m - 1 and j == n - 1:
                return min_remove
            
            visited[i][j] = min_remove
            
            for di, dj in ((0, 1), (1, 0), (0, -1), (-1, 0)):
                new_i, new_j = i + di, j + dj
                if is_valid(new_i, new_j):
                    heappush(pq, (grid[new_i][new_j] + min_remove, new_i, new_j))
        
        return -1

File: 2290-minimum-obstacle-removal-to-reach-corner/2290-minimum-obstacle-removal-to-reach-corner.cpp
class Solution {
private:
    int m, n;
    
    bool isValid(int x, int y){
        return 0 <= x && x < m && 0 <= y && y < n;
    }
    
public:
    int minimumObstacles(vector<vector<int>>& grid) {
        this -> m = grid.size(), this -> n = grid[0].size();
        
        vector visited(m, vector<int>(n, std::numeric_limits<int>::max()));
        priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, greater<tuple<int, int, int>>> pq;
        pq.push({0, 0, 0});
        
        while (!pq.empty()){
            auto [minRemove, i, j] = pq.top(); pq.pop();
            if (minRemove >= visited[i][j]) continue;
            
            if (i == m - 1 && j == n - 1) return minRemove;
            
            visited[i][j] = minRemove;
            for (auto& [di, dj] : vector<pair<int, int>> {{0, 1}, {0, -1}, {1, 0}, {-1, 0}}){
                int newI = i + di, newJ = j + dj;
                if (isValid(newI, newJ))
                    pq.push({grid[newI][newJ] + minRemove, newI, newJ});
                
            }
        }
        
        return -1;
    }
};